### PR TITLE
Add a command to navigate to a certain page by incremental search

### PR DESCRIPTION
**Pull request by valuex**
## IS YOUR COMMIT SIGNED?
Note that this repo requires all commits to be properly signed, [see here](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) for more info
Yes
## Enhancement or Defect Addressed by This PR
- Navigate to a certain page by incremental search
- Move the current page after a selected page

## Description of Proposed Changes
- a pop-up window with a texbox and HirarchyTreeview
- when user input some characters in the textbox, the input chars will be incrementally indexed and the results will be shown in the tree view
- to search Pages title with Chinese characters by inputting the first letters of corresponding Chinese PinYin is also supported
- **Search and NavigateTo**
- search syntax " Section_name  Page_name".   
    - the first _space_ is taken as a seperator, the chars before the _space_ will be used to search Section Name, while the chars after the _space_ will be used to search Page Name.
    -  if there is a leading _space_ input into the textbox, e.g. " onemore", only the pages's name will be searched
    -  **NavigateTo**: click the underlined page name in the tree view
 - **Move to**: one can move the _current page_ after the _destination page_, which the user checked in the tree view

### CODE REVIEW

File: OneMore/Commands/Navigator/NaviDialog.Designer.cs
Comment (stevencohn) [inline OneMore/Commands/Navigator/NaviDialog.Designer.cs:1]: This file is unused. Please delete

File: OneMore/Commands/Navigator/NaviDialog.cs
Comment (stevencohn) [inline OneMore/Commands/Navigator/NaviDialog.cs:1]: This file is unused. Please delete

File: OneMore/Commands/Navigator/NaviDialog.resx
Comment (stevencohn) [inline OneMore/Commands/Navigator/NaviDialog.resx:1]: This file is unused. Please delete

File: OneMore/Commands/Navigator/NaviPagesCommand.cs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+                    //using var one = new OneNote();
+                    //one.SelectLocation(Resx.SearchQF_Title, desc, OneNote.Scope.Sections, Callback);
+
+                    try
+                    {
+                        using var one = new OneNote();
+                        //var service = new SearchServices(owner, one, sectionId);
+
+                        if (copying)
```
Comment (stevencohn) [inline OneMore/Commands/Navigator/NaviPagesCommand.cs:62]: Are these operations going to be implemented? If not, remove the code and associated UI controls from the dialog

File: OneMore/Commands/Search/SearchDialog.resx
Comment (stevencohn) [inline OneMore/Commands/Search/SearchDialog.resx:1]: Is there a reason this file was touched? If not, please revert the changes.

File: OneMore/OneNote.cs [hunk 2]
```diff
@@ -195,13 +195,30 @@ public void Dispose()
...
 		/// <summary>
-		/// Gets or sets whether exceptions are allowed to fall through back to caller or
-		/// are caught and reported by this class. This is a special case for some consumers
-		/// who wish to handle certain exceptions themselves to serve data management.
+		/// Gets all the page structure
 		/// </summary>
-		public bool FallThrough { get; set; }
+		//public string NotebookXml => GetAllPages(); //onenote.Windows.CurrentWindow?.CurrentNotebookId;
+
+        public XElement GetAllPages()
```
Comment (stevencohn) [inline OneMore/OneNote.cs:203]: Should be moved down alphabetically amongst the methods below

File: OneMore/OneNote.cs [hunk 3]
```diff
@@ -195,13 +195,30 @@ public void Dispose()
...
+        {
+            onenote.GetHierarchy(null, HierarchyScope.hsPages, out string xml, XMLSchema.xs2013);
+            if (!string.IsNullOrEmpty(xml))
+            {
```
Comment (stevencohn) [inline OneMore/OneNote.cs:207]: There's not a lot of value in this helper function beyond what GetHierarchy already provides. I'd suggest just inlining it within your business logic rather than adding to the OneNote class.

File: OneMore/OneNote.cs [hunk 4]
```diff
@@ -1303,18 +1320,57 @@ public async Task NavigateTo(string pageId, string objectId)
...
-		public XElement Search(string nodeId, string query, bool unindexed = false)
+        /// <summary>
+        /// Forces OneNote to move a page after one another existed page.
+        /// </summary>
+        /// <param name="CurPageId">Current page ID</param>
+        /// <param name="DestPageId">Destination page ID</param>
+        /// <returns></returns>
+        public async Task MovedAfterPage(string CurPageId, string DestPageId)
+        {
+            var root = GetAllPages();
```
Comment (stevencohn) [inline OneMore/OneNote.cs:1331]: This is very inefficient. There are easier ways of discovering, tracking, and using the notebook and section IDs along with the page and target IDs to make this faster.

File: OneMore/Properties/Resources.ar-SA.resx [hunk 5]
```diff
@@ -3446,6 +3446,10 @@ ISO-code then comma then language name</comment>
     <value>بحث ونسخ / نقل</value>
     <comment>ribbon item</comment>
   </data>
+  <data name="ribNaviButton_Label" xml:space="preserve">
+    <value>FastNavi</value>
```
Comment (stevencohn) [inline OneMore/Properties/Resources.ar-SA.resx:3450]: Is there a better name than "FastNavi"?

File: OneMore/Commands/Navigator/NaviPages.cs [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+                if(strKeyWords=="")
+                {
+                    FilteredPageXml = AllPageXML;
+                }
+                else
+                {
+                    FilteredPageXml = FilterXml(AllPageXML, strKeyWords);
+                }
+
+                hierarchyView11.Nodes.Clear();
```
Comment (stevencohn) [inline OneMore/Commands/Navigator/NaviPages.cs:199]: This is a poor use of TreeView, causing flickering and slowness with each keystroke. You also don't need to fetch AllPages on every keystroke since it is unlikely the hierarchy would change while this dialog is open.

**Review by stevencohn — CHANGES_REQUESTED** [review]

**Comment by stevencohn** [comment]
Honestly, I'm trying to understand where this really belongs. It's functionally equivalent to Search/SearchDialog but with type-ahead, which I'm not sure is enough of a reason for a new dialog.

What do you see as the primary difference?

File: OneMore/Commands/Navigator/NaviPages.cs (hunk 6 above)
Comment (valuex) [reply OneMore/Commands/Navigator/NaviPages.cs:199]: > This is a poor use of TreeView, causing flickering and slowness with each keystroke. You also don't need to fetch AllPages on every keystroke since it is unlikely the hierarchy would change while this dialog is open.

It seems that this variable **AllPageXML** is passed by reference, not by value.
If the value is not fetched very loop, its value will be changed by the function **FilterXML**.
I've tried to define a class with a copy element, but still no luck here.
As for the flickering, I trid serveral methods, also no luck.
Could you give any comments on how to solve those two problems? Thanks.

File: OneMore/Properties/Resources.ar-SA.resx (hunk 5 above)
Comment (valuex) [reply OneMore/Properties/Resources.ar-SA.resx:3450]: Will fix it. How about "Incremental Search"?

**Comment by stevencohn** [comment]
Closing this as I believe there to be very limited value beyond the built-in OneNote `Ctrl+E` search feature which does incremental search on page titles and shows them grouped, such as

image
